python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING, Any, Iterable, Tuple
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf.core.dtypes import CategoricalDtype
+
+if TYPE_CHECKING:
+    from cudf.core.column import CategoricalColumn, ColumnBase
+    from cudf.core.frame import Frame
+
+
+class _Indexer:
+    # Indexer into a column (either a data column or index level).
+    #
+    # >>> df
+    #    a
+    # b
+    # 4  1
+    # 5  2
+    # 6  3
+    # >>> _Indexer("a", column=True).get(df)  # returns column "a" of df
+    # >>> _Indexer("b", index=True).get(df)  # returns index level "b" of df
+
+    def __init__(self, name: Any, column=False, index=False):
+        if column and index:
+            raise ValueError("Cannot specify both column and index")
+        self.name = name
+        self.column, self.index = column, index
+
+    def get(self, obj: Frame) -> ColumnBase:
+        # get the column from `obj`
+        if self.column:
+            return obj._data[self.name]
+        else:
+            if obj._index is not None:
+                return obj._index._data[self.name]
```

Comment:
What if we have a MultiIndex?

Reply:
Synced offline. To clarify for posterity: I think the most general example of when things can go wrong is below:

```python
In [38]: df = cudf.DataFrame({0: [1, 2, 3, 4, 5], 1: [1, 2, 3,  4, 5], 'c': [1, 2, 3, 4, 5]}).set_index([0, 1])
In [39]: df.index.names = ['a', 'b'] # doesn't actually change df._index._data
In [40]: df.merge(df, left_index=True, right_index=True) # KeyError 'a'
In [41]: df.to_pandas().merge(df.to_pandas(), left_index=True, right_index=True)  # Pandas does this no problem
Out[41]:
     c_x  c_y
a b
1 1    1    1
2 2    2    2
3 3    3    3
4 4    4    4
5 5    5    5
```

@kkraus14 what do you think? It feels a bit like a blocker to me. But can fix this afternoon if needed.

Reply:
Yea that feels like a blocker to me as well that should be addressed before merging (assuming we used to handle this properly)

Reply:
I pushed a fix in commit https://github.com/rapidsai/cudf/pull/7454/commits/1cb944804d5dd4d54066d2547a4f69da151228bb.  It's janky, but MultiIndex columns are janky.